xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MultiUserChatServiceImpl.java

```diff
@@ -389,28 +389,18 @@ public void processPacket(final Packet packet) {
         // service. This means that, for instance, a disco request should be responded by the
         // service itself instead of relying on the server to handle the request.
         try {
+            // Check for 'ghost' users (OF-910 / OF-2209 / OF-2369)
+            if (isDeliveryRelatedErrorResponse(packet)) {
+                Log.info("Received a stanza that contained a delivery-related error response from {}. This is indicative of a 'ghost' user. Removing this user from all chat rooms.", packet.getFrom());
+                removeChatUser(packet.getFrom());
+                return;
+            }
             // Check if the packet is a disco request or a packet with namespace iq:register
             if (packet instanceof IQ) {
                 if (process((IQ)packet)) {
                     Log.trace( "Done processing IQ stanza." );
                     return;
                 }
-            } else if (packet instanceof Message) {
```

Comment:
This removal sort of scares ignorant me.  Is the thought that this was previously unreachable code?

Reply:
No, it was reachable. It should now have been replaced with the invocation of `isDeliveryRelatedErrorResponse(packet)` a few lines higher.

Reply:
Thanks, but doesn't that code only handle stanzas of type `IQ` ?

Reply:
That's not the intention. About half-way through that method there's a comment that reads "The remainder of this implementation applies to Message and Presence stanzas."